utils/lmap.ml

```diff
@@ -0,0 +1,84 @@
+module type Thing = sig
+  type t
+
+  val equal : t -> t -> bool
+
+  val print : Format.formatter -> t -> unit
+end
+
+module type S = sig
+  type key
+
+  type (+'a) t
+
+  val empty : 'a t
+  val is_empty : 'a t -> bool
+  val add : key -> 'a -> 'a t -> 'a t
+  val singleton : key -> 'a -> 'a t
+  val disjoint_union : 'a t -> 'a t -> 'a t
+  val disjoint_union_many : 'a t list -> 'a t
+  val iter : (key -> 'a -> unit) -> 'a t -> unit
+  val fold : (key -> 'a -> 'b -> 'b) -> 'a t -> 'b -> 'b
+  val filter : (key -> 'a -> bool) -> 'a t -> 'a t
+  val keys : _ t -> key list
+  val data : 'a t -> 'a list
+  val bindings : 'a t -> (key * 'a) list
+  val of_list : (key * 'a) list -> 'a t
+  val find : key -> 'a t -> 'a
+  val find_opt : key -> 'a t -> 'a option
+  val get_singleton : 'a t -> (key * 'a) option
+  val get_singleton_exn : 'a t -> key * 'a
+  val map: ('a -> 'b) -> 'a t -> 'b t
+  val mapi: (key -> 'a -> 'b) -> 'a t -> 'b t
+  val map_sharing: ('a -> 'a) -> 'a t -> 'a t
+  val filter_map: 'a t -> f:(key -> 'a -> 'b option) -> 'b t
+  val to_seq : 'a t -> (key * 'a) Seq.t
+  val add_seq : (key * 'a) Seq.t -> 'a t -> 'a t
+  val of_seq : (key * 'a) Seq.t -> 'a t
+
+  val print :
+    (Format.formatter -> 'a -> unit) -> Format.formatter -> 'a t -> unit
+end
+
+module Make (T : Thing) : S with type key = T.t = struct
+  type key = T.t
+
+  type (+'a) t = (key * 'a) list
+
+  let empty = []
+  let is_empty m = m = []
+  let add k v m = (k, v) :: m
```

Comment:
This is different from `Map.add`: if you do `Lmap.fold f (Lmap.add x 0 (Lmap.add x 1 Lmap.empty))`, you will call `f` twice instead of once if using `Map`.

Reply:
Okay, changed it to remove the old value.